### PR TITLE
Notifications on actions

### PR DESCRIPTION
**Pull request by captncraig**
By default we will notify all notifications in the crit or warn notifications for the relevant alert keys. We will group notifications so each notification gets executed once per user action.

If an alert specifies `runOnAction = false` it will not be executed on actions.

### CODE REVIEW

**Comment by kylebrandt** [comment]
I'd vote for default on.

**Comment by kylebrandt** [comment]
Also, haven't tested this yet, but are automated actions excluded (i.e autoclosed by bosun)

**Comment by kylebrandt** [comment]
Not sure I understand the config:
```
[kbrandt@grove: ~/src/bosun.org/cmd/bosun] go generate && go run main.go -c kdev.conf -dev -q -r -w -dev
2015/05/14 17:04:03 conf: kdev.conf:57:0: at <actionNotification =...>: notification not found: notification
```
I have `actionNotification = notification` on a line by itself in the config....

**Comment by captncraig** [comment]
@kylebrandt it expects the name of a notification there. My fault for poorly wording that.

**Comment by captncraig** [comment]
Yes, only actions through the web interface get notifications. Auto-close won't fire it.

File: cmd/bosun/conf/conf.go [hunk 1]
```diff
@@ -27,47 +27,47 @@ import (
...
+	TimeAndDate        []int // timeanddate.com cities list
+	ResponseLimit      int64
+	SearchSince        opentsdb.Duration
+	UnknownTemplate    *Template
+	UnknownThreshold   int
+	Templates          map[string]*Template
+	Alerts             map[string]*Alert
+	OrderedAlerts      []*Alert                 `json:"-"` //alerts in order they appear.
+	Notifications      map[string]*Notification `json:"-"`
+	ActionNotification *Notification
```
Comment (mjibson) [inline cmd/bosun/conf/conf.go]: Why not `*Notifications`?

File: cmd/bosun/conf/conf.go [hunk 2]
```diff
@@ -27,47 +27,47 @@ import (
...
+	BlockedPutIPs      []*net.IPNet
+	ShortURLKey        string
 
 	TSDBHost             string                    // OpenTSDB relay and query destination: ny-devtsdb04:4242
 	GraphiteHost         string                    // Graphite query host: foo.bar.baz
 	LogstashElasticHosts expr.LogstashElasticHosts // CSV Elastic Hosts (All part of the same cluster) that stores logstash documents, i.e http://ny-elastic01:9200
 
-	tree            *parse.Tree
-	node            parse.Node
-	unknownTemplate string
```
Comment (mjibson) [inline cmd/bosun/conf/conf.go]: The removal of this has nothing to do with this feature. Can it go in a separate PR?

File: cmd/bosun/sched/notification_test.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	"bytes"
+	"encoding/json"
+	"testing"
+)
+
+func TestActionNotificationTemplates(t *testing.T) {
+	data := &actionNotificationData{}
+
+	err := json.Unmarshal([]byte(`{
+  "States": [
```
Comment (mjibson) [inline cmd/bosun/sched/notification_test.go]: Can you convert this to Go-style instead of JSON?

File: cmd/bosun/sched/notification_test.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	buf := &bytes.Buffer{}
+	err = actionNotificationBodyTemplate.Execute(buf, data)
+	if err != nil {
+		t.Fatal(err)
+	}
+	//fmt.Println(string(buf.Bytes()))
```
Comment (mjibson) [inline cmd/bosun/sched/notification_test.go]: remove

File: cmd/bosun/sched/notification_test.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	buf = &bytes.Buffer{}
+	err = actionNotificationSubjectTemplate.Execute(buf, data)
+	if err != nil {
+		t.Fatal(err)
+	}
+	//fmt.Println(string(buf.Bytes()))
```
Comment (mjibson) [inline cmd/bosun/sched/notification_test.go]: remove

File: cmd/bosun/sched/notify.go [hunk 6]
```diff
@@ -209,3 +211,51 @@ func (s *Schedule) AddNotification(ak expr.AlertKey, n *conf.Notification, start
...
+{{if gt $count 1}} {{$count}} Alerts. 
+{{else}} Incident #{{$first.Last.IncidentId}} ({{$first.Subject}}) 
+{{end}}`
+	body := `{{$count := len .States}}{{.User}} {{.ActionType}} {{$count}} alert{{if gt $count 1}}s{{end}}:
+<strong>Message:</strong> {{.Message}}
+<strong>Incidents:</strong>
+<ul>{{range .States}}
+<li><a href="/incident?id={{.Last.IncidentId}}">#{{.Last.IncidentId}}:<a/> {{.Subject}}</li>{{end}}
+</ul>`
+	actionNotificationSubjectTemplate = template.Must(template.New("").Parse(strings.Replace(subject, "\n", "", -1)))
```
Comment (mjibson) [inline cmd/bosun/sched/notify.go]: Can this use a similar pattern for the unknownTemplate, which has a default but can be set by a user?

File: cmd/bosun/sched/notify.go [hunk 7]
```diff
@@ -209,3 +211,51 @@ func (s *Schedule) AddNotification(ak expr.AlertKey, n *conf.Notification, start
 	}
 	s.Notifications[ak][n.Name] = started
 }
+
+type actionNotificationData struct {
+	States     []*State
+	User       string
+	Message    string
+	ActionType string
```
Comment (mjibson) [inline cmd/bosun/sched/notify.go]: Could this remain of type ActionType?

File: cmd/bosun/sched/notify.go [hunk 8]
```diff
@@ -209,3 +211,51 @@ func (s *Schedule) AddNotification(ak expr.AlertKey, n *conf.Notification, start
...
+	incidents := []*State{}
+	for _, ak := range aks {
+		state := s.GetStatus(ak)
+		if state != nil {
+			incidents = append(incidents, state)
+		}
+	}
+	data := actionNotificationData{incidents, user, message, at.String()}
+	js, err := json.Marshal(data)
+	fmt.Println(string(js))
```
Comment (mjibson) [inline cmd/bosun/sched/notify.go]: remove

File: cmd/bosun/sched/notify.go [hunk 9]
```diff
@@ -209,3 +211,51 @@ func (s *Schedule) AddNotification(ak expr.AlertKey, n *conf.Notification, start
...
+	buf := &bytes.Buffer{}
+	err = actionNotificationSubjectTemplate.Execute(buf, data)
+	if err != nil {
+		return
+	}
+	s.Conf.ActionNotification.Notify(string(buf.Bytes()), string(buf.Bytes()), buf.Bytes(), buf.Bytes(), s.Conf, string(aks[0]))
```
Comment (mjibson) [inline cmd/bosun/sched/notify.go]: `string(buf.Bytes())` -> `buf.String()`

Comment (mjibson) [inline cmd/bosun/sched/notify.go]: `aks[0]` -> `"alertNotification"`

**Comment by kylebrandt** [comment]
@captncraig @gbrayut and I chatted some more. I suggested that we continue with the plan of notifying any notification that corresponds to warn/crit. However, we include a notification level directive that says "No Action emails". So if someone has a contact for say SMS, they can exclude those within the notification directive.

File: cmd/bosun/sched/notify.go (hunk 6 above)
Comment (captncraig) [inline cmd/bosun/sched/notify.go]: That is a good idea.

File: cmd/bosun/sched/notify.go (hunk 7 above)
Comment (captncraig) [inline cmd/bosun/sched/notify.go]: I was having trouble stringing it in the template. That should be possible though.

File: cmd/bosun/sched/notify.go [hunk 10]
```diff
@@ -224,3 +225,87 @@ func (s *Schedule) AddNotification(ak expr.AlertKey, n *conf.Notification, start
...
+var actionNotificationBodyTemplate *htemplate.Template
+
+func init() {
+	subject := `{{$first := index .States 0}}{{$count := len .States}}
+{{.User}} {{.ActionType}}
+{{if gt $count 1}} {{$count}} Alerts. 
+{{else}} Incident #{{$first.Last.IncidentId}} ({{$first.Subject}}) 
+{{end}}`
+	body := `{{$count := len .States}}{{.User}} {{.ActionType}} {{$count}} alert{{if gt $count 1}}s{{end}}:
+<strong>Message:</strong> {{.Message}}
```
Comment (mjibson) [inline cmd/bosun/sched/notify.go]: need a new block after this

File: cmd/bosun/sched/notify.go [hunk 11]
```diff
@@ -224,3 +225,87 @@ func (s *Schedule) AddNotification(ak expr.AlertKey, n *conf.Notification, start
...
+<strong>Incidents:</strong>
+<ul>{{range .States}}
```
Comment (mjibson) [inline cmd/bosun/sched/notify.go]: could the whitespace here be fixed?

```
<ul>
{{range .States}}
  <li>
    <a href="/incident?id={{.Last.IncidentId}}">#{{.Last.IncidentId}}:<a/>
    {{.Subject}}
  </li>
{{end}}
</ul>
```

File: cmd/bosun/conf/conf.go [hunk 12]
```diff
@@ -995,6 +997,8 @@ func (c *Conf) loadNotification(s *parse.SectionNode) {
 				c.error(err)
 			}
 			n.Body = tmpl
+		case "runOnActions":
```
Comment (mjibson) [inline cmd/bosun/conf/conf.go]: Why not a global setting? That'd be consistent with our other settings.

File: cmd/bosun/sched/template.go [hunk 13]
```diff
@@ -433,3 +423,18 @@ func (c *Context) LSQueryAll(index_root, keystring, filter, sduration, eduration
...
+	User       string
+	Message    string
+	ActionType ActionType
+
+	schedule *Schedule
+}
+
+func (a actionNotificationContext) IncidentLink(i uint64) string {
+	return a.schedule.Conf.MakeLink("/incident", &url.Values{
+		"id": []string{fmt.Sprintf("%d", i)},
```
Comment (mjibson) [inline cmd/bosun/sched/template.go]: `fmt.Sprint(i)` is sufficient

File: cmd/bosun/conf/conf.go (hunk 12 above)
Comment (captncraig) [inline cmd/bosun/conf/conf.go]: The use case here is that I want to email on actions, but maybe we don't want chat or some remote api or whatever to get those. This allows specific notifications to be exempt from action notifications.

File: cmd/bosun/sched/notify.go [hunk 14]
```diff
@@ -224,3 +225,85 @@ func (s *Schedule) AddNotification(ak expr.AlertKey, n *conf.Notification, start
...
+{{if gt $count 1}} {{$count}} Alerts. 
+{{else}} Incident #{{$first.Last.IncidentId}} ({{$first.Subject}}) 
+{{end}}`
+	body := `{{$count := len .States}}{{.User}} {{.ActionType}} {{$count}} alert{{if gt $count 1}}s{{end}}: <br/>
+<strong>Message:</strong> {{.Message}} <br/>
+<strong>Incidents:</strong> <br/>
+<ul>
+	{{range .States}}
+		<li>
+			<a href="{{$.IncidentLink .Last.IncidentId}}">#{{.Last.IncidentId}}:<a/>
```
Comment (mjibson) [inline cmd/bosun/sched/notify.go]: `/a`

File: cmd/bosun/sched/notify.go [hunk 15]
```diff
@@ -224,3 +225,85 @@ func (s *Schedule) AddNotification(ak expr.AlertKey, n *conf.Notification, start
...
+	actionNotificationBodyTemplate = htemplate.Must(htemplate.New("").Parse(body))
+}
+
+func (s *Schedule) ActionNotify(at ActionType, user, message string, aks []expr.AlertKey) {
+	groupings := s.groupActionNotifications(aks)
+
+	for notification, aks := range groupings {
+		incidents := []*State{}
+		for _, ak := range aks {
+			state := s.GetStatus(ak)
```
Comment (mjibson) [inline cmd/bosun/sched/notify.go]: What if `groupActionNotifications` returned a `map[*conf.Notification][]*State`? This would prevent calling GetStatus twice for each ak.

File: cmd/bosun/sched/notify.go [hunk 16]
```diff
@@ -224,3 +225,85 @@ func (s *Schedule) AddNotification(ak expr.AlertKey, n *conf.Notification, start
...
+			if state != nil {
+				incidents = append(incidents, state)
+			}
+		}
+		data := actionNotificationContext{incidents, user, message, at, s}
+
+		buf := &bytes.Buffer{}
+		err := actionNotificationSubjectTemplate.Execute(buf, data)
+		if err != nil {
+			continue
```
Comment (mjibson) [inline cmd/bosun/sched/notify.go]: Could we at least log this error?

File: cmd/bosun/sched/template.go [hunk 17]
```diff
@@ -433,3 +423,18 @@ func (c *Context) LSQueryAll(index_root, keystring, filter, sduration, eduration
...
+type actionNotificationContext struct {
+	States     []*State
+	User       string
+	Message    string
+	ActionType ActionType
+
+	schedule *Schedule
+}
+
+func (a actionNotificationContext) IncidentLink(i uint64) string {
```
Comment (mjibson) [inline cmd/bosun/sched/template.go]: probably `*actionNotificationContext`

**Comment by mjibson** [comment]
Minor changes. Otherwise LGTM.
